Add database

MaXM91 opened the pull request.

MaXM91 committed:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml.

Commit:
style : править оформление

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : удалить неиспользуемый импорт и не нужные комментарии.

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

Commit:
style : править оформление

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -3,17 +3,16 @@
-    private Set<Integer> usersLike = new HashSet<>();
     private int id;
 
     @NotBlank(message = "film: name is blank!")
```
NotBlank включает в себя NotEmpty

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -25,21 +24,27 @@ public class Film {
 
-    public void setUsersLike(Integer id) {
-        usersLike.add(id);
+    public Film(int id, String name, String description, List<Genre> genres, Mpa mpa, Integer rate,
```
Вместо шаблонного кода можно использовать lombok

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -5,16 +5,15 @@
     private int id;
 
-    private Set<Integer> userFriends = new HashSet<>();
+    private HashMap<Integer, Boolean> userFriends = new HashMap<>();
```
Для чего нужен Map вместо Set?

MaXM91 replied:
В предыдущей реализации интерфейса вроде как надо было просто держать список добавленных друзей, в этой реализации необходимо хранить подтверждение дружбы. Что должно быть в объекте не указанно, просто друзья или друзья подтвержденные, или все друзья которых добавил со статусом подтверждения - не указано. Верну обратно на SET, буду загружать только друзей, которых добавил. Посмотрю как пройдут тесты...

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -2,29 +2,33 @@
+        this.likeService = likeService;
     }
 
     public Film create(Film film) throws ValidationException {
```
throws для Runtime исключений избыточен. Аналогично для остальных методов

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+        }
+    }
+
+    private List<Genre> changeFilmGenre(Film film) {
```
Можно проще:
1) избавиться от выполнения contains
2) избавиться от множественных запросов в БД

Например, 
    private List<Genre> changeFilmGenre(Film film) {
        List<Genre> uniqueGenres = film.getGenres().stream()
                .distinct()
                .collect(Collectors.toList());

        jdbcTemplate.batchUpdate(
                "INSERT INTO film_genre (film_id, genre_id) VALUES(?,?)",
                new BatchPreparedStatementSetter() {
                    public void setValues(PreparedStatement ps, int i) throws SQLException {
                        ps.setLong(1, film.getId());
                        ps.setLong(2, uniqueGenres.get(i).getId());
                    }

                    public int getBatchSize() {
                        return uniqueGenres.size();
                    }
                });

        return jdbcTemplate.query("SELECT * FROM genre g join film_genre fg on g.genre_id = fg.genre_id WHERE fg.film_id = ?",
                (rs, rowNum) -> new Genre(rs.getInt("genre_id"), rs.getString("name")),
                film.getId());
    }

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+                    rs.getString("description"), readFilmGenre(rs.getInt("id")),
+                    readFilmMpa(rs.getInt("id")), readUsersLike(id),
+                    rs.getDate("releaseDate").toLocalDate(), rs.getInt("duration")), id);
+        } catch (RuntimeException exc) {
```
Перехватывать RuntimeException - плохая практика. То есть несмотря на любую поломку программы, считаем, что просто не найден пользователь, что неверно. 
Лучше конкретизировать тип исключения, который может быть брошен методом queryForObject.
Аналогично для остальных методов

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    @Override
+    public List<Film> get() {
```
По названию get непонятно, что он возвращает)

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+
+        return jdbcTemplate.query("SELECT * FROM films",
+            (rs, rowNum) -> new Film(rs.getInt("id"), rs.getString("name"),
+                rs.getString("description"), readFilmGenre(rs.getInt("id")),
```
Методы readFilmGenre и readUsersLike создают проблему N+1 запроса.
Пусть всего в БД находится 1 000 000 фильмов.
Запросом SELECT * FROM films получим все 1 000 000 фильмов за одно обращение к БД. 
А чтобы заполнить жанры и лайки, нужно выполнить еще 1 000 000 запросов для жанров и 1 000 000 запросов для лайков.
Получим 2 000 000 + 1 запросов в БД. А хотим ограничиться только 1 + 2 запросами )
Аналогично для popularFilms

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+                    rs.getDate("releaseDate").toLocalDate(), rs.getInt("duration")), (int) count);
+
+        } catch (EmptyResultDataAccessException e) {
+            return null;
```
null лучше не возвращать, если есть такая возможность. Можно заменить на пустой список.
Аналогично для остальных методов

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+    private List<Genre> readFilmGenre(Integer id) {
+        try {
+            List<Genre> genres = jdbcTemplate.query(
+                "SELECT fg.genre_id, g.name \n" + "FROM films AS f\n" +
```
Использование films здесь лишнее

MaXM91 replied:
метод удален

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,210 @@
+                    "LEFT JOIN film_genre AS fg ON f.id=fg.film_id \n" +
+                    "LEFT JOIN genre AS g ON fg.genre_id=g.genre_id\n" + "WHERE f.id = ?\n" +
+                    "GROUP BY g.genre_id", (rs, rowNum) -> {
+                    if (rs.getInt("genre_id") == 0) {
```
0 в каком случае будет?

MaXM91 replied:
Где-то читал, при отсутствии данных-> int - 0, String - null.

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/genre/GenreDbStorage.java:
```diff
@@ -0,0 +1,30 @@
+        try {
+            return jdbcTemplate.queryForObject("SELECT * FROM genre WHERE genre_id = ?",
+                (rs, rowNum) -> new Genre(rs.getInt("genre_id"), rs.getString("name")), id);
+        } catch (RuntimeException exc) {
```
RuntimeException лучше заменить на более детализированное исключение. 
При формировании ObjectNotFoundException лучше указать id, по которому ничего не найдено

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/like/LikeDbStorage.java:
```diff
@@ -0,0 +1,21 @@
+public class LikeDbStorage {
+    private final JdbcTemplate jdbcTemplate;
+
+    public LikeDbStorage(JdbcTemplate jdbcTemplate) {
```
Шаблонный код можно заменить lombok-ом. Аналогично в остальных классах

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/mpa/MpaDbStorage.java:
```diff
@@ -0,0 +1,29 @@
+        try {
+            return jdbcTemplate.queryForObject("SELECT * FROM mpa WHERE mpa_id = ?",
+                (rs, rowNum) -> new Mpa(rs.getInt("mpa_id"), rs.getString("name")), id);
+        } catch (RuntimeException ecx) {
```
RuntimeException лучше заменить на более детализированное исключение. 
При формировании ObjectNotFoundException лучше указать id, по которому ничего не найдено

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+            List<Integer> friends = foundFriendsIds(id);
+
+            for (Integer friendId : friends) {
+                assert user != null;
```
Для чего assert в каждой итерации, при том, что ссылка на user не меняется?

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+
+            for (Integer friendId : friends) {
+                assert user != null;
+                user.setUserFriends(friendId);
```
Можно установить всех друзей без цикла?

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+                rs.getDate("birthday")
+                    .toLocalDate());
+
+            List<Integer> friends = foundFriendsIds(rs.getInt("id"));
```
Здесь проблема N+1 запроса аналогично жанрам и лайкам

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+        List<User> friends = new ArrayList<>();
+        List<Integer> friendsIds = foundFriendsIds(id);
+
+        for (Integer idFriend : friendsIds) {
```
Лучше получить всех друзей одним запросом вместо использования цикла

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+
+    @Override
+    public void addFriend(Integer id, Integer idFriend) {
+        jdbcTemplate.update("INSERT INTO friends (user_id_from, user_id_to) VALUES(?,?)", id, idFriend);
```
А если отправить 1 000 000 запросов подряд на добавление юзером 1 юзера 2, как будет обработана ошибка БД?

MaXM91 replied:
Добавлен контроль выбрасываемого методом update исключения

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+        jdbcTemplate.update("INSERT INTO friends (user_id_from, user_id_to) VALUES(?,?)", id, idFriend);
+
+        if (checkFriends(id, idFriend)) {
+            jdbcTemplate.update("UPDATE friends SET approved_by_to = 1 WHERE user_id_from = ? AND user_id_to = ?", id,
```
В чем профит иметь на каждую дружбу по 2 записи и еще статус дружбы? Статус тогда выглядит лишним

MaXM91 replied:
исправлено, статус исключен.

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+        List<User> mutualFriends = new ArrayList<>();
+
+        for (User user : friendsId) {
+            if (friendsOtherId.contains(user)) {
```
Пересечение друзей можно попробовать сделать на уровне запроса через джойны

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,167 @@
+    private boolean checkFriends(Integer id, Integer idFriend) {
+        List<User> checkFriend = getFriends(idFriend);
+
+        for (User user : checkFriend) {
```
Проверку можно сделать на уровне запроса

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/validators/FilmCombinedSerializer.java:
```diff
@@ -0,0 +1,48 @@
+import java.io.IOException;
+
+@JsonComponent
+public class FilmCombinedSerializer {
```
Закомментированный код лишний

MaXM91 replied:
удалено

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml.

Commit:
style : править оформление

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml.

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml.

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml.

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/model/ObjectFilm.java:
```diff
@@ -0,0 +1,14 @@
+
+@Data
+@AllArgsConstructor
+public class ObjectFilm {
```
Не используется

MaXM91 replied:
удалено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -2,112 +2,119 @@
     }
 
-    public Film delete(Integer id) throws ObjectNotFoundException {
+    public Film delete(Integer id) {
```
лучше добавить валидацию в контроллер, избавившись от лишнего  кода,
 добавив Positive перед id
Пример
    @DeleteMapping("/{id}")
    public Film deleteFilm(@PathVariable @Positive Integer id) {
        return filmService.delete(id);
    }
И еще на уровень контроллера нужно добавить Validated, чтобы эти аннотации обработались.
Аналогично для остальных методов и сервисов

MaXM91 replied:
При введении валидации, SBV отправляет 400 ошибку. На примере удаления лайка с отрицательным ид пользователя, наблюдаем требование к обработке и определению как 404 ошибки.

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -2,112 +2,119 @@
+            throw new ValidationException("update film: film have bad id - " + film.getId() + "!");
+        }
+
         if (filmStorage.found(film.getId()) == null) {
```
Довольно много дублирующегося кода вида
if (  == null) {
throw new Exception(...);
}
Лучше вынести повторяющийся код в отдельные методы

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -37,7 +36,7 @@ public Film getFilm(@PathVariable Integer id) {
+        return filmService.getFilms();
     }
 
     @PutMapping("/{id}/like/{userId}")
```
Лучше использовать напрямую likeService для добавления и удаления лайка

MaXM91 replied:
введено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -19,133 +17,129 @@ public class UserService {
-
+    public User found(Integer id) {
         User user = userStorage.found(id);
         if (user == null) {
```
Довольно много дублирующегося кода вида
if ( == null) {
throw new Exception(...);
}
Лучше вынести повторяющийся код в отдельные методы

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,265 @@
+                    "FROM films AS f\n" + "LEFT JOIN likes AS l ON f.ID = l.FILM_ID\n" + "WHERE f.id = ?" +
+                    "GROUP BY f.id",
+                (rs, rowNum) -> new Film(rs.getInt("id"), rs.getString("name"), rs.getString("description"),
+                    FilmDbStorage.this.readGenre(rs.getInt("id")), FilmDbStorage.this.readMpa(rs.getInt("id")),
```
Использование имени класса (FilmDbStorage....readeGenres) для нестатических методов избыточно

NikolaiDem replied:
FilmDbStorage.this.readGenre(rs.getInt("id")) получение жанров создает проблему N+1 запроса, так как появляется необходимость для каждого фильма идти в БД. Необходимо получить все жанры за один запрос в БД сразу для всех фильмов

NikolaiDem replied:
mpa можно добавить непосредственно в sql запрос, чтобы снова не ходить в БД

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,265 @@
+                    genre = new Genre(rs.getInt("genre_id"), rs.getString("genre_name"));
+                    genres.add(genre);
+                    films.get(rs.getInt("id")).setGenres(genres);
+                    rs.next();
```
rs.next() можно вынести за пределы if/else

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/genre/GenreDbStorage.java:
```diff
@@ -0,0 +1,47 @@
+public class GenreDbStorage {
+    private final JdbcTemplate jdbcTemplate;
+
+    public Genre found(Integer id) {
```
Методы очень сильно дублируются

MaXM91 replied:
Вынесен общий private метод

NikolaiDem replied:
Можно сделать намного проще:
    public Genre found(Integer id) {
        List<Genre> genres = jdbcTemplate.query("SELECT * FROM genre where genre_id = ?", getGenreRowMapper(), id);
        if (genres.isEmpty()) {
            throw new ObjectNotFoundException(////////);
        }
        return genres.get(0);
    }

    public List<Genre> get() {
        return jdbcTemplate.query("SELECT * FROM genre", getGenreRowMapper());
    }

    private RowMapper<Genre> getGenreRowMapper() {
        return (rs, rowNum) -> new Genre(rs.getInt("genre_id"), rs.getString("genre_name"));
    }

Аналогично для mpa

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/like/LikeDbStorage.java:
```diff
@@ -0,0 +1,43 @@
+public class LikeDbStorage {
+    private final JdbcTemplate jdbcTemplate;
+
+    public void add(Integer filmId, Integer userId) {
```
Методы очень сильно дублируются

MaXM91 replied:
Вынесен общий private метод

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/mpa/MpaDbStorage.java:
```diff
@@ -0,0 +1,46 @@
+public class MpaDbStorage {
+    private final JdbcTemplate jdbcTemplate;
+
+    public Mpa found(Integer id) {
```
Методы очень сильно дублируются

MaXM91 replied:
Вынесен общий private метод

NikolaiDem commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,73 @@
+    CONSTRAINT uniq_likes_string UNIQUE (film_id, user_id)
+);
+
+CREATE TABLE IF NOT EXISTS film_mpa
```
Какая связь предполагается между film и mpa? Если не many-to-many, то для чего создана отдельная таблица? Можно mpa_id хранить в фильме

MaXM91 replied:
Выполнено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,281 @@
+                    "FROM friends AS f\n" +
+                    "LEFT JOIN users AS u ON f.user_id_to = u.id\n" +
+                    "LEFT JOIN friends ON u.id = friends.user_id_from\n" +
+                    "WHERE f.user_id_from = ? AND f.user_id_to IN (SELECT user_id_to\n" +
```
При использовании 2х джойнов вложенный селект здесь явно лишний

MaXM91 replied:
исправлено

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml.

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/model/ObjectFilm.java:
```diff
@@ -0,0 +1,13 @@
+public class ObjectFilm {
+    private int id;
+
+    public ObjectFilm(int id) {
```
исправлено

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/model/Mpa.java:
```diff
@@ -0,0 +1,14 @@
+    private Integer id;
+    private String name;
+
+    public Mpa(int id, String name) {
```
исправлено

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/model/Genre.java:
```diff
@@ -0,0 +1,14 @@
+    private Integer id;
+    private String name;
+
+    public Genre(Integer id, String name) {
```
исправлено

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml.

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml;
- актуализировать схему(schem.png).

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml;
- актуализировать схему(schem.png).

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml;
- актуализировать схему(schem.png).

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -2,112 +2,75 @@
-        if (userService.found(userId) == null) {
-            log.info("При удалении лайка не найден юзер с ид {}", userId);
-            throw new ObjectNotFoundException("film/removeLike: user not found!");
+    private void checkIds(Integer id, String methodName) {
```
Для определения места возникновения исключения можно использовать StackTrace, передавать метод в исключение избыточно.
Аналогично для остальных сервисов

MaXM91 replied:
исправлено

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/like/LikeService.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private void checkIds(Integer filmId, Integer userId, String methodName) {
+        if (filmId <= 0) {
```
Так как по требованиям нужно возвращать на отрицательный id 404 ошибку, это значит, что отрицательные id являются валидными, тогда проверки на 34 и 39 строках являются лишними

MaXM91 replied:
Исправлено

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/like/LikeDbStorage.java:
```diff
@@ -0,0 +1,37 @@
+    private void changeLikes(String sql, Integer filmId, Integer userId, String type) {
+        try {
+            jdbcTemplate.update(sql, filmId, userId);
+        } catch (EmptyResultDataAccessException exc) {
```
Вместо всевозможных перехватов исключений лучше удостовериться, что они не возникнут, выполнив предварительные проверки.
Аналогично для остальных репозиториев

MaXM91 replied:
Исправлено. Отлавливание исключений оставлено на поисковых методах. (found(Integer id)), все остальные методы кроме getAll, getFriends и подобных проверятся методами found на сервисном уровне.

MaXM91 replied:
исправлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,240 @@
+                "SELECT f.user_id_to, u.email, u.login, u.name, u.birthday, fr.user_id_to\n" +
+                    "FROM friends AS f\n" +
+                    "LEFT JOIN users AS u ON f.user_id_to = u.id\n" +
+                    "LEFT JOIN friends AS fr ON u.id = fr.user_id_from \n" +
```
Для чего здесь 2-ой джойн?

MaXM91 replied:
2 джойн для того чтобы заполнить SET<Integer> с идами друзей у найденных друзей. В предыдущем модуле по тз заполняли список друзей. Тут реализую тоже. Находим простым запросом друзей пользователя ид, первым джойном получаем их данные, вторым джойном список ид их друзей. По полученным данным формируем List<User> друзей с полностью заполненной, необходимой информацией. Также выполняется в других запросах, где формируется объект User.

MaXM91 replied:
для добавления ид друзей найденным друзьям

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/service/genre/GenreService.java:
```diff
@@ -0,0 +1,23 @@
+package ru.yandex.practicum.filmorate.service.genre;
```
Исправлено

MaXM91 replied:
исправлено

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,287 @@
+
+            response = createFilmsFromRows(rs);
+
+            if (response.size() == 0) {
```
Мы либо делаем так, либо получаем проблему n+2 запроса. Проблемы выдачи популярных фильмов как таковой нету. Делаем подзапрос фром, там же ставим лимит и собираем строки... Но мы ориентируемся на тесты, а по тестам фильм без лайков(т.е. фильм отсутствующий в таблице) тоже участвует в рейтинге, и тут простого запроса не получается, либо использовать то, что мы еще не проходили, или преобразовать столбцы в строку и все также писать дополнительный код для обработки. Выполнил так чтобы не было дополнительных запросов к БД и использовал уже вынесенную часть кода с других методов для формирования списка объектов.

MaXM91 replied:
Изменил sql запрос, тепрь этого нету

MaXM91 replied:
исправлен запрос sql

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,287 @@
+    public List<Film> popularFilms(long count) {
+        try {
+            List<Film> response;
+            int num = (int) count * 2;
```
Переформировал sql запрос. Теперь все нормально

MaXM91 replied:
исправлен sql запрос

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,287 @@
+            } else {
+                return new ArrayList<>(response.subList(0, (int) count));
+            }
+        } catch (ArrayIndexOutOfBoundsException exp) {
```
исправлено

MaXM91 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    @Override
+    public List<Film> get() {
```
исправлено

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml;
- актуализировать схему(schem.png).

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database

Commit:
feat :
- добавить функционал для работы с ДБ (MpaDbStorage.class, GenreDbStorage, LikeDbStorage, UserDbStorage.class, FilmDbStorage.class);
- добавить загрузочные данные для проведения внутренних тестов (schema.sql, data.sql);
- добавить внутренние тесты с покрытием каждого метода (FilmorateApplicationTest.class);
- править оформление на соответствие checkstyle.xml;
- актуализировать схему(schem.png).

Commit:
Merge remote-tracking branch 'origin/add-database' into add-database
